### DISCUSSION TITLE
Preview feature feedback: `@prisma/adapter-planetscale` / PlanetScale Serverless Driver support

### DISCUSSION THREAD

**Discussion by janpio** [prisma/prisma#21347] [Feedback]
Support for the **[PlanetScale Serverless Driver](https://github.com/planetscale/database-js)** is in preview behind the `driverAdapters` preview feature. 

Please share your feedback about `@prisma/adapter-planetscale` released in [v5.4.0](https://github.com/prisma/prisma/releases/tag/5.4.0) in this issue.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repository.
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this discussion.
- If you have any questions, don't hesitate to ask them in the [Prisma Slack](https://slack.prisma.io/) community

**Comment by Murtatrxx** [comment]
Finally

**Comment by JClackett** [comment]
Can I ask what the purpose of this package is/what does it solve? just wondering if it's useful for me. I'm currently using planetscale and have a Remix app running on vercel, to combat the cold start times, im using the Accelerate console and using the @prisma/client/edge import. Would this replace the need to use the Accelerate console?

**Reply by janpio** [reply]
Good question! 

From the Prisma ORM side, this new package just makes it possible for you to use https://github.com/planetscale/database-js (the PlanetScale serverless driver) to talk to your PlanetScale database, additionally to our built in MySQL driver that we have been using for years. Some people want or need that, so we made it possible.

PlanetScale has built that package to solve a similar use case to what [Prisma Accelerate](https://www.prisma.io/data-platform/accelerate) solves. Their announcement blog post at https://planetscale.com/blog/introducing-the-planetscale-serverless-driver-for-javascript or the tutorial at https://planetscale.com/docs/tutorials/planetscale-serverless-driver explains that a bit more. 

If you are happy with Prisma Accelerate, and my colleagues from that team sure hope you are, there is no reason for you to migrate to the PlanetScale Serverless Driver and our new driver adapter `@prisma/adapter-planetscale`. Prisma Accelerate also has additional functionality like caching, and is of course very tightly and deeply integrated with Prisma ORM - so this should be the optimal combination.

But now you have the _option_ to choose the thing that works best in your situation. 

Temporarily replacing the native driver with this one should be pretty simple, see the tiny code changes documented in the release notes. Maybe just try it out, run your test suite, and see how it works for you.

**Reply by JClackett** [reply]
Thanks for the response! I think a concern I had with accelerate was having to manage multiple platforms for my database connection and, whilst prototyping a product I don't really want to have to pay at that point (I believe accelerate will be a paid platform soon right?). But I can definitely see the caching + insights adding a lot of value as a product evolves passed the prototype phase!

Something i'm still unsure of when using the new adapter: would I still need to use the @prisma/client/edge import somehow with the --no-engine generation and the DIRECT_URL + DATABASE_URL connections? (im still deploying to serverless so really need to keep bundle size down and cold starts to a minimum). However if I keep the --no-engine it says it requires the prisma:// url. or is the whole point of this adapter to make it so that we dont have to use the --no-engine flag + /edge import?

**Reply by janpio** [reply]
> would I still need to use the @prisma/client/edge import somehow with the --no-engine generation and the DIRECT_URL + DATABASE_URL connections? (im still deploying to serverless so really need to keep bundle size down and cold starts to a minimum). However if I keep the --no-engine it says it requires the prisma:// url. or is the whole point of this adapter to make it so that we dont have to use the --no-engine flag + /edge import?

Good question. I need to investigate that a bit - will get back to you.

**Reply by JClackett** [reply]
Yeah I did quick test, and if using the edge import + no-engine I have to provide it with a prisma:// url, otherwise it throws an error. But then i'm concerned that ill be importing the full prisma library again in my serverless functions, but when using that I also have a runtime [issue](https://github.com/prisma/prisma/issues/21369) too - which is being looked at!

**Reply by janpio** [reply]
Driver Adapters currently still need and use the full query engine, as you would use when you used the native driver and talked to the database directly. So just using `@prisma/client/edge` can not work yet. We are still working on improving and evolving this situation, with the goal of support Edge Function deployments as well for example. That will most probably also include ways to make the size of the required package smaller. 

If you have concrete problems caused by the usage of the current package, please do open an issue by the way - this should not cause any trouble.

**Reply by JClackett** [reply]
Okay, good to know! in that case maybe i'm missing the point of the planetscale adapter? what's the difference if I just pass the planetscale db url into prisma or pass it via the adapter? just that it uses HTTP instead of TCP?

**Reply by janpio** [reply]
Yes, and that it talks to the PlanetScale equivalent of Prisma's Accelerate. 
Before, using the PlanetScale serverless driver was not possible with Prisma, now it is.

**Reply by JClackett** [reply]
Ah yeah, so we get the benefits of the serverless connection pooling via planetscale, just not edge ready prisma client yet. Will see if the cold starts become a problem again and if so may move back to accelerate

**Reply by rwieruch** [reply]
@JClackett thanks for your questions and @janpio thanks for all the responses! I'd curious if you have seen any performance improvements regarding cold starts @JClackett ? I am in a similar position where I am using Next + Vercel Edge Runtime + Primsa/Edge/Client + Data Proxy (predecessor (?) of Accelerate) to get Prisma running on the edge.

**Reply by JClackett** [reply]
@rwieruch Haven't fully quantified it yet tbh but in general I do feel like it's a lot better using the planetscale driver to manage connections compared to the regular prisma connection mechanism. Maybe slightly worse cold start times compared to using accelerate + edge client, but having to manage two database urls (DATABASE_URL + DIRECT_URL) + the pain of developing with a local database when using the edge client has made be stick with this adapter for now! I've even bailed on using vercel edge runtime because it wasn't that much quicker for me, plus deploy times went from 1min to 7min!

**Comment by Alekzv9** [comment]
FYI: if you are using `pscale cli` to connect 

`pscale connect DB BRANCH`
`DATABASE_URL='mysql://127.0.0.1:3306 ...`

It won't work. You need to use the full connection string

**Reply by jln13x** [reply]
Is there no workaround?

**Reply by janpio** [reply]
Workaround for what exactly? If I understood @Alekzv9 he posted a workaround: use the full connection string.

**Reply by Alekzv9** [reply]
Yes, this is the workaround. 

If I try to use the planet scale cli it throws the following error


```
library: 'SSL routines',
    function: 'ssl3_get_record',
    reason: 'wrong version number',
    code: 'ERR_SSL_WRONG_VERSION_NUMBER'
   ```

**Reply by janpio** [reply]
I think the planetscale CLI might not provide the endpoint for the PlanetScale serverless driver, it needs a proxy in front of the normal database. You might want to ask PlanetScale what their process for local testing is.

**Reply by mattrobenolt** [reply]
Hi there, for local testing, I've poked around with https://github.com/mattrobenolt/ps-http-sim but haven't gotten much feedback so it's been a bit difficult to finish this off and get it in more hands, but it's intended to work with a local mysqld running on your local.

I am also working on allowing HTTP connections in `pscale connect`, so that's also WIP for the meantime.

As mentioned, with this being HTTP and not TCP, it's brought some challenges getting this exposed in all the right places. Likely this will be enabled with a flag to `pscale connect --http=8080` or something to enable it, but nothing as of yet.

**Reply by dani-mp** [reply]
Hey, folks. I'm experiencing this right now and found this thread. Has there been any progress on this? The dx of connecting to the different branches with the PlanetScale CLI and then doing `npx prisma studio` is pretty neat, wouldn't enjoy losing that.

**Comment by shiyuhang0** [comment]
Is there any mechanism to add custom adapters with `driverAdapters` preview feature?

**Reply by janpio** [reply]
Not a documented and supported (as in: we guarantee we do not just break it between versions) one. But the adapters of course are just Typescript and JavaScript code and use an API that is publicly accessible, so in theory you can look at [how the current adapters are implemented](https://github.com/prisma/prisma-engines/tree/main/query-engine/driver-adapters/js), write your own variant, and use it as `adapter` when creating a `PrismaClient` instance.

(We might make this an official API with a stable surface later after we have stabilized it via our own adapters.)

What database driver are you interested in writing an adapter for?

**Reply by shiyuhang0** [reply]
I work for TiDB Cloud and want to add [TiDB Cloud serverless driver](https://docs.pingcap.com/tidbcloud/serverless-driver).
It's okay to add adapters by Prisma official at the beginning. But I think contributing adapters by database vendors will be a more suitable way if it is possiable, and we are willing to contribute ours when the adapter mechanism is built.

**Reply by Jolg42** [reply]
For context https://github.com/prisma/prisma-engines/pull/4342

**Reply by ceddybi** [reply]
@janpio @Jolg42 let's say you want to create a custom adapter for a database like Couchbase, does it require to write custom adapters in both rust and typescript?

**Comment by hmbrg** [comment]
Does using this serverless driver mean I can use Prisma in any serverless/edge environment today?

**Reply by JClackett** [reply]
No, all this does is change from TCP to HTTP when connecting to the database. If you want it to run on the edge, you'll have to use prisma accelerate (paid soon). This will work with regular nodejs serverless though, just with a worse cold start time!

**Reply by janpio** [reply]
It's a bit more nuanced @JClackett. 

TLDR: Not yet, but soon.

Prisma can already be used in **serverless environments** like AWS Lambda or Vercel Serverless Functions with your normal database driver, and now with Driver Adapters like this on. [Prisma Accelerate](https://www.prisma.io/data-platform/accelerate) can be a solution for when you hit problems caused by the nature of how serverless scales (= many, many connection pools at the same time).

For **Edge Function environments** like Cloudflare Workers or Vercel Edge Functions, _currently_ you can only use Prisma via Prisma Accelerate and `@prisma/client/edge`. The release of Driver Adapters is the first step to bring us to being able to allow usage of Prisma Client directly. Our classic database drivers use TCP, which is not available or just via some experimental approaches - so we needed to develop support for HTTP based drivers first (like this one from PlanetScale). Now we can focus on making it possible to run Prisma Client directly there, using these Driver Adapters.

Follow this issue for updates: https://github.com/prisma/prisma/issues/21394 And if you have not, please fill this survey and let us know which deployment providers your are most interested in: https://pris.ly/survey/driver-adapters-github

**Comment by kylegwalsh** [comment]
Based on your comment @janpio, this adapter does not allow us to remove any custom Lambda layer logic for AWS yet?

> Driver Adapters currently still need and use the full query engine, as you would use when you used the native driver and talked to the database directly

As an example, this project shows how a layer needed to be created in order to leverage Prisma in a Lambda environment previously: https://github.com/jetbridge/sst-prisma/blob/master/stacks/layers.ts

**Reply by janpio** [reply]
The example you linked shows how a Lambda layer _can_ be used. You can totally deploy Prisma Client to AWS Lambda without the use of a separate Lambda layer.

But the query engine, which the layer separates from the rest of the application mainly, is still required, yes.

**Comment by michaelhays** [comment]
Troubleshooting tip for those who are having an issue with connecting to the database: Check the database host you're using. If it's in the format of `abcd1234.us-east-2.psdb.cloud`, try [creating a new database password](https://planetscale.com/docs/concepts/password-roles#create-a-password-with-custom-role) and check that the host is instead `aws.connect.psdb.cloud`.

So for clarity, your `DATABASE_URL` should look like this:

```
mysql://user:password@aws.connect.psdb.cloud/database?sslaccept=strict
```

not this:

```
mysql://user:password@abcd1234.us-east-2.psdb.cloud/database?sslaccept=strict
```

Credit to [this comment](https://github.com/prisma/prisma/issues/21364#issuecomment-1747956175) for pointing me in the right direction.

**Comment by glenntws** [comment]
Awesome, any updates on plans to make the use of Prisma Edge Client compatible with Planetscale drivers as well? Would allow us to switch over to edge functions and saving quite a bit of money + much faster API response times. Prisma Accelerate is nice but I don't see the need for such a tool in the specific scenario Prisma + Planetscale HTTP driver, especially since the pricing is not yet disclosed.

**Reply by janpio** [reply]
Yes: https://github.com/prisma/prisma/issues/21394 
Support for serverless drivers was a prerequisite for that.

**Comment by JClackett** [comment]
There seems to be a type error now in 5.5.0:

```ts
import { connect } from "@planetscale/database"
import { PrismaPlanetScale } from "@prisma/adapter-planetscale"
import { PrismaClient } from "@prisma/client"

const connection = connect({ url: process.env.DATABASE_URL })
const adapter = new PrismaPlanetScale(connection)

export const prisma =  new PrismaClient({ adapter }) // error here <----
```


```
Type 'PrismaPlanetScale' is not assignable to type 'DriverAdapter'.
  The types returned by 'startTransaction()' are incompatible between these types.
    Type 'Promise<Result<Transaction>>' is not assignable to type 'Promise<Result_3<Transaction_2>>'.
      Type 'Result<Transaction>' is not assignable to type 'Result_3<Transaction_2>'.
        Type '{ map<U>(fn: (value: Transaction) => U): Result<U>; flatMap<U>(fn: (value: Transaction) => Result<U>): Result<U>; } & { ...; }' is not assignable to type 'Result_3<Transaction_2>'.
          Type '{ map<U>(fn: (value: Transaction) => U): Result<U>; flatMap<U>(fn: (value: Transaction) => Result<U>): Result<U>; } & { ...; }' is not assignable to type '{ map<U>(fn: (value: Transaction_2) => U): Result_3<U>; flatMap<U>(fn: (value: Transaction_2) => Result_3<U>): Result_3<U>; } & { ...; }'.
            Type '{ map<U>(fn: (value: Transaction) => U): Result<U>; flatMap<U>(fn: (value: Transaction) => Result<U>): Result<U>; } & { ...; }' is not assignable to type '{ map<U>(fn: (value: Transaction_2) => U): Result_3<U>; flatMap<U>(fn: (value: Transaction_2) => Result_3<U>): Result_3<U>; }'.
              Types of property 'map' are incompatible.
                Type '<U>(fn: (value: Transaction) => U) => Result<U>' is not assignable to type '<U>(fn: (value: Transaction_2) => U) => Result_3<U>'.
                  Types of parameters 'fn' and 'fn' are incompatible.
                    Types of parameters 'value' and 'value' are incompatible.
                      Type 'Transaction' is not assignable to type 'Transaction_2'.

// and so on....

```

all 5.5.0

**Reply by Jolg42** [reply]
@JClackett What's the version for all your Prisma packages? So `prisma`, `@prisma/client`, `@prisma/adapter-planetscale`? And also for `@planetscale/database`?

**Reply by Arechii** [reply]
Same issue here; packages `prisma`, `@prisma/client` and `@prisma/adapter-planetscale` are on v5.5.0. `@planetscale/database` is v1.11.0.

**Reply by JClackett** [reply]
yeah same as above, 5.5.0 for everything, 1.11.0 planetscale

**Reply by Jolg42** [reply]
Thanks for the details, we'll definitely investigate this tomorrow, for now, as a workaround you can downgrade to the `5.4.2` version for Prisma packages.

**Reply by SevInf** [reply]
Can confirm the issue on our side - https://github.com/prisma/prisma/issues/21613
We are working on a fix. Workarounds meanwhile are either downgrading as @Jolg42 mentioned in a previous comment or `@ts-ignore` on offending line — adapter actually works in runtime, only types are the problem.

**Reply by SevInf** [reply]
@JClackett @Arechii could you try version `5.5.1-dev.1` and see if that fixes the issue for you? This a dev snapshot, it fixed the problem on my reproduction. We do not recommend using dev snapshots in production but if the fix works for you, it will be published as a stable `5.5.1` version soon after.

**Reply by Arechii** [reply]
Thanks for the quick work, I can confirm that 5.5.1-dev.1 fixes it for me!

**Reply by Jolg42** [reply]
@SevInf just did the official patch release, you can now use https://github.com/prisma/prisma/releases/tag/5.5.1

**Comment by ruheni** [comment]
If you upgrade to Prisma v5.6.0 and run into the following error:
```
 [TypeError]: PrismaPlanetScale must be initialized with an instance of Client.
```

... you can get around it using the following snippet:
```ts
import { createRequire } from "node:module";
import { PrismaPlanetScale } from "@prisma/adapter-planetscale";

const require = createRequire(import.meta.url);
const { Client } = require("@planetscale/database");

const client = new Client({ url:  process.env.DATABASE_URL  })
const adapter = new PrismaPlanetScale(client);
```

This will be resolved in the next release.

**Reply by JClackett** [reply]
Any updates for when this release will be published?

**Reply by Jolg42** [reply]
@JClackett `5.7.0` is planned for next Tuesday, December 5th.

**Reply by janpio** [reply]
And it's out @JClackett.

**Reply by JClackett** [reply]
What is the recommended approach to using the adapter now? do we still have to use the "createRequire" function?

I still get:
```ts
const connection = connect({ url: env.DATABASE_URL })
const adapter = new PrismaPlanetScale(connection)

// Argument of type 'Connection' is not assignable to parameter of type 'Client'.
// Property 'connection' is missing in type 'Connection' but required in type 'Client'.ts(2345)
```

5.7.0 versions for everything, 1.11.0 for planetscale

**Reply by JClackett** [reply]
Just figured it, it should be:
```ts
import { Client } from "@planetscale/database"
import { PrismaPlanetScale } from "@prisma/adapter-planetscale"
import { PrismaClient } from "@prisma/client"


const client = new Client({ url: process.env.DATABASE_URL })
const adapter = new PrismaPlanetScale(client)
```

Maybe worth updating the original discussion to show this as it still uses the old "connection" from planetscale.

**Reply by janpio** [reply]
Where exactly @JClackett? Happy to, but could not easily find it.

**Reply by JClackett** [reply]
@janpio https://github.com/prisma/prisma/releases/tag/5.4.0

This is the only place I've seen the usage mentioned

**Reply by SevInf** [reply]
@JClackett `createRequire` was a temporary workaround for regression in `5.6.0` in ESM file. Any earlier or newer version, as well as common.js module, does not need it.

**Reply by JClackett** [reply]
@SevInf yep, am aware - just saying that I couldn't find any official docs for setting up the adapter based on the new updates - would be worth mentioning

**Reply by janpio** [reply]
Yes, we are going to modify both the release notes and blog post to reflect the current state somehow.
It's complicated when we change stuff that quickly 😆

**Comment by bdematt** [comment]
I'm seeing the error `Error: PrismaClient is unable to run in Cloudflare Workers. As an alternative, try Accelerate: https://pris.ly/d/accelerate.`  (in a Cloudflare Worker, obviously, running through Wrangler)

I followed the steps for PlanetScale Driver in 5.7.0 and later in the [release notes](https://github.com/prisma/prisma/releases/tag/5.4.0)

```typescript
import { Hono } from 'hono'
import { Client } from '@planetscale/database';
import { PrismaPlanetScale } from '@prisma/adapter-planetscale';
import { PrismaClient } from '@prisma/client';

const app = new Hono()

app.get('/', async (c) => {
  const client = new Client({ url: `${c?.env?.DATABASE_URL}` });
  const adapter = new PrismaPlanetScale(client);
  const prisma = new PrismaClient({ adapter });
  const users = await prisma.user.findMany(); // Throws here
  return c.json(users)
})

export default app
```

**Reply by Jolg42** [reply]
For **Edge Function environments** like Cloudflare Workers or Vercel Edge Functions, _currently_ you can only use Prisma via Prisma Accelerate and `@prisma/client/edge`. 

Follow this issue for updates: https://github.com/prisma/prisma/issues/21394 

We’re working on bringing Edge function support to Prisma ORM and we would appreciate your input by [submitting a response to our survey](https://pris.ly/survey/edge-functions-release-notes). By filling out the survey, you will be considered for our Early Access cohort as soon as we have something for you to try out!

**Reply by bdematt** [reply]
Cool thanks, good to know it's on its way 👍
